### PR TITLE
Add column to image kernel

### PR DESCRIPTION
**Pull request by bartekxk**

### CODE REVIEW

File: include/ck/tensor_operation/gpu/device/device_conv_tensor_rearrange.hpp [hunk 1]
```diff
@@ -12,21 +12,24 @@ namespace tensor_operation {
...
 
 /**
- * \brief Image to column.
+ * \brief Convolution Tensor Rearrange.
  *
- * This Device operator converts image ([G, N, Di, Hi, Wi, C]) to the gemm
- * problem([N * Do * Ho * Wo, Z *  Y * X * C]). G must be equal to 1.
+ * This Device operator supports conversion image ([G, N, Di, Hi, Wi, C]) to
+ * the gemm problem([N * Do * Ho * Wo, Z *  Y * X * C]) (Image to Column) and
+ * conversion gemm form to the image (Column to Image). G must be equal to 1.
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/device_conv_tensor_rearrange.hpp]:
```suggestion
 * conversion gemm form to the image (Column to Image). 
 
 Note that G must be equal to 1.
```

Comment (aosewski) [reply include/ck/tensor_operation/gpu/device/device_conv_tensor_rearrange.hpp]: Leave this note about `G` in separate line. Just to be more visible.

File: library/include/ck/library/tensor_operation_instance/gpu/conv_tensor_rearrange/device_column_to_image_instance.hpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        DeviceColumnToImageImpl<NDimSpatial, InLayout,       BF16,        BF16,    64,    64,    64,   S<8, 8>,     8, ColumnToImage>,
+        DeviceColumnToImageImpl<NDimSpatial, InLayout,       BF16,        BF16,   128,    16,    16,  S<8, 16>,     1, ColumnToImage>,
+        DeviceColumnToImageImpl<NDimSpatial, InLayout,       BF16,        BF16,   128,    64,    64,  S<8, 16>,     1, ColumnToImage>,
+        DeviceColumnToImageImpl<NDimSpatial, InLayout,       BF16,        BF16,   128,    32,    64,  S<8, 16>,     4, ColumnToImage>,
+        DeviceColumnToImageImpl<NDimSpatial, InLayout,       BF16,        BF16,   128,    64,   128,  S<8, 16>,     8, ColumnToImage>,
+        DeviceColumnToImageImpl<NDimSpatial, InLayout,       BF16,        BF16,   256,    16,    16, S<16, 16>,     1, ColumnToImage>,
+        DeviceColumnToImageImpl<NDimSpatial, InLayout,       BF16,        BF16,   256,    64,    64, S<16, 16>,     1, ColumnToImage>,
+        DeviceColumnToImageImpl<NDimSpatial, InLayout,       BF16,        BF16,   256,   128,   128, S<16, 16>,     1, ColumnToImage>,
```
Comment (aosewski) [inline library/include/ck/library/tensor_operation_instance/gpu/conv_tensor_rearrange/device_column_to_image_instance.hpp]: Those instances aim to be generic one right? Is it needed to have that many of them? (Also for all other block sizes)

File: library/src/tensor_operation_instance/gpu/column_to_image/device_column_to_image_nhwc_1d_instance.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace ck {
+namespace tensor_operation {
+namespace device {
+namespace instance {
+
+using namespace ck::conv_tensor_rearrange_op;
+
+#ifdef CK_ENABLE_BF16
+void add_device_column_to_image_nhwc_1d_bf16_instances(
```
Comment (aosewski) [inline library/src/tensor_operation_instance/gpu/column_to_image/device_column_to_image_nhwc_1d_instance.cpp]: Shouldn't layout name in function name reflect the used one ? The function name contain 2d layout and itself is for 1D. This is confusing.

File: include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp [hunk 4]
```diff
@@ -5,64 +5,40 @@
...
           typename OutputDataType,
           index_t BlockSize,
           index_t MPerBlock,
           index_t KPerBlock,
           typename ThreadClusterLengths,
-          index_t ScalarPerVector>
-struct DeviceImageToColumnImpl
-    : public DeviceImageToColumn<NDimSpatial, InputLayout, InputDataType, OutputDataType>
+          index_t ScalarPerVector,
+          typename ConvTensorRearrangeOp>
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp]: Is it necessary here ? The DeviceOp name already indicates it should be ImageToColumn operation, so you can pass appropriate enumerator to the base class.

File: include/ck/tensor_operation/gpu/device/impl/device_column_to_image_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+#include "ck/host_utility/io.hpp"
+
+namespace ck {
+namespace tensor_operation {
+namespace device {
+
+// Image to column for input layout NDHWC:
+//   input : image converted to the gemm problem [N * Do * Ho * Wo, Z * Y * X * C]
+//   output : image [N, Di, Hi, Wi, C]
+template <index_t NDimSpatial,
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_column_to_image_impl.hpp:29]: You can constrain this to supported values using `enable_if` here.

**Review by aosewski — COMMENTED** [review]
I've made some comments. I haven't reviewed whole `device_column_to_image_impl.hpp` file yet. Will do it next week.

File: include/ck/tensor_operation/gpu/device/device_conv_tensor_rearrange.hpp (hunk 1 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/device_conv_tensor_rearrange.hpp]: Done.

File: library/src/tensor_operation_instance/gpu/column_to_image/device_column_to_image_nhwc_1d_instance.cpp (hunk 3 above)
Comment (bartekxk) [reply library/src/tensor_operation_instance/gpu/column_to_image/device_column_to_image_nhwc_1d_instance.cpp]: Done.

File: library/include/ck/library/tensor_operation_instance/gpu/conv_tensor_rearrange/device_column_to_image_instance.hpp (hunk 2 above)
Comment (bartekxk) [reply library/include/ck/library/tensor_operation_instance/gpu/conv_tensor_rearrange/device_column_to_image_instance.hpp]: Removed.

File: include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp (hunk 4 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp]: Removed.

File: include/ck/tensor_operation/gpu/device/impl/device_column_to_image_impl.hpp (hunk 5 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_column_to_image_impl.hpp:29]: Done.

**Review by aosewski — DISMISSED** [review]

**Review by saadrahim — DISMISSED** [review]
Please update the changelog.md file if appropriate.

I've also noticed a few copyright years off on newly created files but that is not documentation related.

File: include/ck/tensor_operation/gpu/device/conv_tensor_rearrange_op.hpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: MIT
```
Comment (saadrahim) [inline include/ck/tensor_operation/gpu/device/conv_tensor_rearrange_op.hpp:1]: appears to be new file with copyright starting in the past.

Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/conv_tensor_rearrange_op.hpp:1]: Changed.

**Review by saadrahim — APPROVED** [review]

**Pull request merged**